go/client/chat_api_handler.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package client
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/keybase/client/go/protocol/chat1"
+	"golang.org/x/net/context"
+)
+
+// ErrInvalidOptions is returned when the options aren't valid.
+type ErrInvalidOptions struct {
+	method  string
+	version int
+	err     error
+}
+
+func (e ErrInvalidOptions) Error() string {
+	return fmt.Sprintf("invalid %s v%d options: %s", e.method, e.version, e.err)
+}
+
+// Call represents a JSON chat call.
+type Call struct {
+	Jsonrpc string
+	ID      int
+	Method  string
+	Params  Params
+}
+
+// Params represents the `params` portion of the JSON chat call.
+type Params struct {
+	Version int
+	Options json.RawMessage
+}
+
+// CallError is the result when there is an error.
+type CallError struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+// Reply is returned with the results of procressing a Call.
+type Reply struct {
+	Jsonrpc string      `json:"jsonrpc,omitempty"`
+	ID      int         `json:"id,omitempty"`
+	Error   *CallError  `json:"error,omitempty"`
+	Result  interface{} `json:"result,omitempty"`
+}
+
+// ChatAPIHandler can handle all of the chat json api methods.
+type ChatAPIHandler interface {
+	ListV1(context.Context, Call, io.Writer) error
+	ReadV1(context.Context, Call, io.Writer) error
+	SendV1(context.Context, Call, io.Writer) error
+}
+
+// ChatServiceHandler can call the service.
+type ChatServiceHandler interface {
+	ListV1(context.Context) Reply
+	ReadV1(context.Context, readOptionsV1) Reply
+	SendV1(context.Context, sendOptionsV1) Reply
+}
+
+// ChatAPI implements ChatAPIHandler and contains a ChatServiceHandler
+// to do all the work.
+type ChatAPI struct {
+	svcHandler ChatServiceHandler
+	indent     bool
+}
+
+// ChatChannel represents a channel through which chat happens.
+type ChatChannel struct {
+	Name      string `json:"name"`
+	Public    bool   `json:"public"`
+	TopicType string `json:"topic_type,omitempty"`
+	TopicName string `json:"topic_name,omitempty"`
+}
+
+// Valid returns true if the ChatChannel has at least a Name.
+func (c ChatChannel) Valid() bool {
+	return len(c.Name) > 0
+}
+
+// TopicTypeEnum returns the chat1.TopicType from the TopicType string field.
+func (c ChatChannel) TopicTypeEnum() chat1.TopicType {
+	if len(c.TopicType) == 0 {
```

Comment:
It's a bit confusing that `0` implies different topic types in `keybase chat api` vs local service. If we want to make topic type optional with default being CHAT, shall we all change the service definition to have the same corresponding values?

(not proposing a change for this PR; just thought I'd bring this up for discussion)